File: problem_004.hs
-- A palindromic number reads the same both ways. The largest palindrome made from
-- the product of two 2-digit numbers is 9009 = 91 × 99.
-- 
-- Find the largest palindrome made from the product of two 3-digit numbers.
-- 
-- http://projecteuler.net/index.php?section=problems&id=4

module Main where

digits :: Integral t => t -> [t]
digits n
  | n < 10 = [n]
  | otherwise = y:digits x 
  where
    (x, y) = divMod n 10

palindrome :: Integral t => t -> Bool
palindrome n = d == (reverse d) where d = digits n

products :: Integral t => [t]
products = [x * y | x <- [100..999], y <- [100..x]]

main :: IO ()
main = do print $ maximum $ filter (palindrome) products

File: ruby/problem_004.rb



class Fixnum
  def digits
    d = []
    n = self
    until n < 10 do
      d << n % 10
      n /= 10
    end
    d << n
    return d.reverse
  end
  
  def palindrome?
    d = self.digits
    d == d.reverse
  end
end

biggest = 0
(100..999).each do |a|
  (100..a).each do |b|
    product = a * b
    biggest = product if product > biggest and product.palindrome?
  end
end
p biggest
